tests/director/tests/application_api_test.go

```diff
@@ -492,34 +494,133 @@ func TestCreateApplicationWithDuplicatedNamesWithinTenant(t *testing.T) {
+
+		//THEN
+		require.NoError(t, err)
+	})
+
+	t.Run("Error when application is in formation", func(t *testing.T) {
+		//GIVEN
+		ctx := context.Background()
+		tenantID := tenant.TestTenants.GetIDByName(t, "TestDeleteApplicationIfInFormation")
+		expectedErrorMsg := "graphql: The operation is not allowed [reason=System deletion failed: the system is part of a formation - test-scenario]"
+
+		defaultValue := "DEFAULT"
+		scenarios := []string{defaultValue, "test-scenario"}
+
+		jsonSchema := map[string]interface{}{
+			"type":        "array",
+			"minItems":    1,
+			"uniqueItems": true,
+			"items": map[string]interface{}{
+				"type": "string",
+				"enum": scenarios,
+			},
+		}
+		var schema interface{} = jsonSchema
+
+		fixtures.CreateLabelDefinitionWithinTenant(t, ctx, dexGraphQLClient, ScenariosLabel, schema, tenantID)
+
+		applicationInput := fixtures.FixSampleApplicationRegisterInput("first")
+		applicationInput.Labels = &graphql.Labels{ScenariosLabel: scenarios}
+		appInputGQL, err := testctx.Tc.Graphqlizer.ApplicationRegisterInputToGQL(applicationInput)
+		require.NoError(t, err)
+
+		createApplicationReq := fixtures.FixRegisterApplicationRequest(appInputGQL)
+		application := graphql.Application{}
+
+		err = testctx.Tc.RunOperationWithCustomTenant(ctx, dexGraphQLClient, tenantID, createApplicationReq, &application)
+		require.NoError(t, err)
+		require.NotEmpty(t, application.ID)
+
+		defer func() {
+			fixtures.UpdateApplicationScenariosToDefaultState(t, ctx, dexGraphQLClient, tenantID, application.ID)
+			fixtures.UnregisterApplication(t, ctx, dexGraphQLClient, tenantID, application.ID)
+		}()
+
+		//WHEN
+		req := fixtures.FixUnregisterApplicationRequest(application.ID)
+
+		err = testctx.Tc.RunOperationWithCustomTenant(ctx, dexGraphQLClient, tenantID, req, nil)
+
+		//THEN
+		require.EqualError(t, err, expectedErrorMsg)
+	})
+
+	t.Run("Error when application is in formation and runtime", func(t *testing.T) {
+		//GIVEN
+		expectedErrorMsg := "graphql: The operation is not allowed [reason=System deletion failed: the system is part of a formation - test-scenario and of runtime - one-runtime]"
+		ctx := context.Background()
+		tenantID := tenant.TestTenants.GetIDByName(t, "TestDeleteApplicationIfInFormation")
+
+		runtimeInput := fixtures.FixRuntimeInput("one-runtime")
+		defaultValue := "DEFAULT"
+		scenarios := []string{defaultValue, "test-scenario"}
+		(*runtimeInput.Labels)[ScenariosLabel] = scenarios
+		runtimeInputWithNormalizationGQL, err := testctx.Tc.Graphqlizer.RuntimeInputToGQL(runtimeInput)
+		require.NoError(t, err)
+		registerRuntimeRequest := fixtures.FixRegisterRuntimeRequest(runtimeInputWithNormalizationGQL)
+
+		runtime := graphql.Runtime{}
+		err = testctx.Tc.RunOperationWithCustomTenant(ctx, dexGraphQLClient, tenantID, registerRuntimeRequest, &runtime)
+		require.NoError(t, err)
+		require.NotEmpty(t, runtime.ID)
+		defer fixtures.UnregisterRuntime(t, ctx, dexGraphQLClient, tenantID, runtime.ID)
```

Comment:
usually you defer first, then do whatever you would be doing. This makes sure defer would be called even if what you had to do failed half way